### DISCUSSION TITLE
How to get groups and projects for current user ?

### DISCUSSION THREAD

**Discussion by dd-ssc** [python-gitlab/python-gitlab#2124] [Q&A] [answered]
## How to get groups and projects for current user ?

I'm not sure if I am doing something wrong, but I try to get my groups and projects on gitlab.com and I see quite a difference:

Mostly taken from the [Getting started > Managers docs](https://python-gitlab.readthedocs.io/en/stable/api-usage.html#managers):

```
from gitlab import Gitlab
gitlab_token = '<snipped>'
gitlab = Gitlab(private_token=gitlab_token)
# this gets me (an iterator over) all MY groups
groups = gitlab.groups.list(iterator=True)
# this gets me (an iterator over) ALL (public ?) projects on gitlab.com
projects = gitlab.projects.list(iterator=True)
```

Looking at https://github.com/python-gitlab/python-gitlab/pull/404#issuecomment-770674223, it seems I need to get projects via some user object, will try that out next. Still wondering...

Is this the intended behavior ? 
Why the difference ?

## Expected Behavior

I would expect `groups.list(...)` and `projects.list(...)` to use the same approach (least surprise principle) and either both return _my_ resources or _all public_ resources, but not do something for the first and something else for the second resource.

## Actual Behavior

+ `groups.list(...)` returns (an iterator over) _my_ groups (i.e. the groups belonging to the currently authenticated GitLab user.
+ `projects.list(...)` returns (an iterator over) _all_ projects (presumably the public ones) on the GitLab instance.

## Specifications

  - python-gitlab version: `3.6.0`
  - API version you are using (v3/v4): `v4`
  - Gitlab server version (or gitlab.com): `gitlab.com`

**Comment by JohnVillalovos** [comment, marked as answer]
To retrieve a user's projects would do something like this:

```
from gitlab import Gitlab
gitlab_token = '<snipped>'
gl = Gitlab(private_token=gitlab_token)
gl.users.get(USER_ID)
for index, project in enumerate(gl.users.projects.list(iterator=True)):
    print(f"{index}: {project!r} {project.web_url}")

```

https://docs.gitlab.com/ee/api/projects.html#list-user-projects

But as the Gitlab docs state this only retrieves projects located under the user's namespace. Not all projects that the user has access to.

Now to get all projects that the current user is a member of can do:
```
for index, project in enumerate(gl.projects.list(membership=True, iterator=True)):
    print(f"{index}: {project!r} {project.web_url}")
```
https://docs.gitlab.com/ee/api/projects.html#list-all-projects

**Comment by nejch** [comment]
Regarding the difference in behavior, yes this seems to just be intended GitLab API behavior. in the Groups API docs:
![image](https://user-images.githubusercontent.com/16777978/177260534-9006c252-fe59-4895-aae7-53efaacce8a6.png)

You can achieve the same for projects by using the User projects API as John said above (if interested in your namespace only) or by passing the `membership` parameter (if interested in all your projects) in your Projects API call:
![image](https://user-images.githubusercontent.com/16777978/177260697-31981ba0-b23d-4503-828e-c0d8fc296ff3.png)

One of the many quirks of GitLab's API - not much we can do here as a client library :)

Edit: ah I see the membership thing was already there in the answer :P
